convert/convert.go

```diff
@@ -61,6 +75,52 @@ func Convert(inputFilename, outputFilename string, from, to Format) error {
 	return nil
 }
 
+func convertKongGateway2xTo3x(input *file.Content, filename string) (*file.Content, error) {
+	if input == nil {
+		return nil, fmt.Errorf("input content is nil")
+	}
+	outputContent := input.DeepCopy()
+
+	convertedRoutes := []*file.FRoute{}
+	for _, service := range outputContent.Services {
+		for _, route := range service.Routes {
+			convertedRoutes = append(convertedRoutes, migrateRoutesPathFieldPre300(route, filename))
+		}
+		service.Routes = convertedRoutes
+	}
+
+	if outputContent.FormatVersion != "3.0" {
+		cprint.UpdatePrintf(
+			"From the '%s' config file, the _format_version field has been migrated from '%s' to '%s'.\n"+
+				"These automatic changes may not be correct or exhaustive enough, please\n"+
+				"perform a manual audit of the config file.\n\n"+
+				"For related information, please visit: https://docs.konghq.com/deck/latest/3.0-upgrade\n",
+			filename, outputContent.FormatVersion, "3.0")
+		outputContent.FormatVersion = "3.0"
+	}
+	return outputContent, nil
+}
+
+func migrateRoutesPathFieldPre300(route *file.FRoute, filename string) *file.FRoute {
+	changedPaths := []string{}
+	for _, path := range route.Paths {
+		if !strings.HasPrefix(*path, "~/") && utils.IsPathRegexLike(*path) {
+			changedPaths = append(changedPaths, *path)
+			*path = "~" + *path
+		}
+	}
+	if len(changedPaths) > 0 {
+		cprint.UpdatePrintf(
+			"From the '%s' config file, a regex pattern usage was detected\n"+
+				"for the '%s' paths in the 'paths' field.\n"+
+				"Kong gateway versions 3.0 and above require that regular expressions\n"+
+				"start with a '~' character to distinguish from simple prefix match.\n"+
+				"In order to make these paths compatible with 3.x, a '~' prefix has been added.\n\n",
+			filename, strings.Join(changedPaths, ", "))
```

Comment:
The same draconian thing exists here. The code shouldn't print this 10k times if there are 10k routes that violate the prefix requirement.

Reply:
https://github.com/Kong/deck/pull/753/commits/6347de40814d87c4ad9d9365b01202b315c640c4